tools/codesigningtool/codesigningtool.py

```diff
@@ -120,16 +123,100 @@ def _certificate_fingerprint(identity):
+
+  return cert
 
+def _find_smart_card_identities(identity=None):
+  """Finds smartcard identitites on the current system."""
+  ids = []
+  # Use 'system_profiler' to export XML
+  # containing smartcards keychain data
+  _, xml, _ = execute.execute_and_filter_output([
+      "/usr/sbin/system_profiler",
+      "SPSmartCardsDataType",
+      "-xml"
+  ], raise_on_failure=True)
+  xml = ET.fromstring(xml)
+  # Find 'AVAIL_SMARTCARDS_TOKEN' node and move up one level
+  # to find the 'array' of available tokens, each one is a 'string' node
+  available_tokens_xml = xml.findall('.//*/string[.="AVAIL_SMARTCARDS_TOKEN"]../array')[0]
+  available_tokens_xml = available_tokens_xml.findall('.//*/string')
+  tokens = [t.text for t in available_tokens_xml]
+  # Find 'AVAIL_SMARTCARDS_KEYCHAIN' node and move up one level
+  # to find the 'array' of available keychain items
+  smartcards_keychain = xml.findall('.//*/string[.="AVAIL_SMARTCARDS_KEYCHAIN"]../array')[0]
+  # For each 'token' decode the certificate and for each non-expired cert:
+  #
+  # 1. Check if 'identity' was provided and if it matches a 'CN', in that case stop the loop
+  #    and return the respective `SHA1`
+  # 2. Otherwise append each `SHA1` found to 'ids' to be returned at the end
+  #
+  # ps: note that if 'identity' is provided and it does not match any existing item in the
+  # smartcard keychain 'ids' will be empty, so this function's behaviour is consistent with
+  # '_find_codesign_identities' where it's being called
+  for token in tokens:
+    # Used to find certificate info
+    cert_start = "-----BEGIN CERTIFICATE-----"
+    # All 'string' nodes at the same level of 'token'
+    all_string_nodes = smartcards_keychain.findall('.//*/string[.="{}"]../string'.format(token))
+    # Find only the 'string' nodes that hold certificate data and are not empty
+    cert_data = [s.text for s in all_string_nodes if cert_start in s.text and s.text.strip() != ""]
+
+    for data in cert_data:
+      data_split = data.split('\n')
+      # Extract expiry date and continue the loop if certificate is expired. The row being processed looks like this:
+      #
+      # Valid from: 2021-02-12 21:35:04 +0000 to: 2022-02-12 21:35:05 +0000, SSL trust: NO, X509 trust: YES
+      #
+      expiry_date_data = [s for s in data_split if "Valid from:" in s]
+      if len(expiry_date_data) == 0:
+        continue
+      expiry_date_data = expiry_date_data[0]
+      expiry_date = re.search(r"(?<=to:)(.*?)(?=,)", expiry_date_data).group().strip()
+      expiry_date = datetime.datetime.strptime(expiry_date, "%Y-%m-%d %H:%M:%S %z")
+      now = datetime.datetime.now(expiry_date.tzinfo)
+      if now > expiry_date:
+        continue
+
+      # This is valid identity, decode the certificate, extract
+      # Common Name and Fingerprint and handle their values accordingly
+      # as described above
+      cert = data_split[data_split.index(cert_start) + 1]
+      cert = base64.b64decode(cert)
```

Comment:
Would it be potentially easier to parse this (and parts earlier around `cert_data` if we instead extracted the cert using `re` and just capturing the string between `-----BEGIN CERTIFICATE-----` and `-----END CERTIFICATE-----`?

Technically I think it's also already in a PEM form which could be passed to `openssl -inform PEM` without needing to b64decode, however it doesn't have the line breaks every 64 characters so that'd have to be added

Reply:
Let me give this a shot, also the code would look cleaner with such regex I think.

Reply:
Using a regex for this has made things mode readable IMO, thx for the suggestion!

Regarding the PEM format comment, I'm reusing an existing function `_certificate_fingerprint` and adding a one liner to decode seems simpler than creating a new function for `-inform PEM` and handling the line breaks just for this?